### PR TITLE
[autoparallel] new metainfoprop based on metainfo class

### PR DESCRIPTION
**Pull request by Cypher30**
# What's New?

In the previous PR, we introduce `MetaInfo` class to process op meta data (memory, flops). In this PR, I introduce a new `MetaInfoProp` based on this mechanism so that we could combine our SPMD solver and activation checkpoint solver.

NOTE: 
1. The content in this PR is experimental, it might contain some bugs that need to be fixed in future PR after we do some tests on this mechanism.
2. For @super-dainiu , you might carefully look at the backward phase and the shape consistency node as you modify the `_construct_chain()` method for the activation checkpoint solver. Currently, all the attached tensors in PR #2162 follow the original `MetaInfoProp` behavior in `colossalai.fx.passes`. But the shape consistency node will discard the backward input after the first comm operation (see #2144 for detail). You also need to carefully modify the temp memory needed in `_construct_chain()`, as I use the `fwd_mem_tmp` and `bwd_mem_tmp` attr in `GraphInfo` class. Contact me if you meet any problems.

### CODE REVIEW

File: colossalai/auto_parallel/passes/meta_info_prop.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            # attach it to graph_info
+            graph_info.fwd_tmp = buffer_tensor
+
+            # set data_ptr for output_tensor
+            for tensor in output_tensor:
+                self._set_data_ptr(tensor)
+            # attach it to graph_info
+            graph_info.fwd_out = output_tensor
+
+        # fetch other memory informations
```
Comment (Cypher30) [inline colossalai/auto_parallel/passes/meta_info_prop.py:153]: @super-dainiu I fetch temp memory informations here and store them in `GraphInfo.fwd_mem_tmp` and `GraphInfo.bwd_mem_tmp`

File: colossalai/auto_parallel/passes/runtime_preparation_pass.py [hunk 2]
```diff
@@ -78,6 +78,10 @@ def _solution_annotatation(gm: torch.fx.GraphModule,
         origin_node_sharding_spec_dict[node_index] = strategies_vector[strategy_index].get_sharding_spec_by_name(
             str(node))
 
+        # attach the corresponding metainfo if node has the attribute `metainfo_vector`
+        if hasattr(node, 'metainfo_vector'):
```
Comment (Cypher30) [inline colossalai/auto_parallel/passes/runtime_preparation_pass.py:82]: @super-dainiu meta info for the best strategy is attached to the node here.

File: colossalai/auto_parallel/passes/meta_info_prop.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    def _set_data_ptr(self, x):
+        """
+        Set uuid to tensor
+        """
+        if isinstance(x, torch.Tensor):
+            if not x.data_ptr():
+                data_ptr = uuid.uuid4()
+                x.data_ptr = lambda: data_ptr
+
+    def _is_inpalce(self, node: Node):
```
Comment (super-dainiu) [inline colossalai/auto_parallel/passes/meta_info_prop.py:40]: typo

File: colossalai/auto_parallel/passes/meta_info_prop.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+    @compatibility(is_backward_compatible=False)
+    def placeholder_handler(self, node: Node) -> None:
+        """
+        Handle the placeholder node.
+        """
+        graph_info = GraphInfo()
+        out = getattr(node, '_meta_data', None)
+        if out is not None:
+            out = [out]
```
Comment (super-dainiu) [inline colossalai/auto_parallel/passes/meta_info_prop.py:66]: better use `normalize_tuple()`
what if `out` is already a tuple of tensors?

Comment (Cypher30) [reply colossalai/auto_parallel/passes/meta_info_prop.py]: fixed

File: colossalai/auto_parallel/passes/meta_info_prop.py (hunk 3 above)
Comment (Cypher30) [reply colossalai/auto_parallel/passes/meta_info_prop.py]: fixed

**Review by super-dainiu — APPROVED** [review]

**Pull request merged**